docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+# ADR 054: Semver Compatible SDK Modules
+
+## Changelog
+
+* 2022-04-27: First draft
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+In order to move the Cosmos SDK to a system of decoupled semantically versioned
+modules which can be composed in different combinations (ex. staking v3 with
+bank v1 and distribution v2), we need to reassess how we organize the API surface
+of modules to avoid problems with go semantic import versioning and
+circular dependencies. This ADR explores various approaches we can take to
+addressing these issues.
```

Comment:
Is it the intent here that users should be able to write a program that's free to compose together arbitrary versions of Cosmos SDK modules — maintained in the cosmos-sdk repository by Cosmos SDK maintainers — while also reliably satisfying the expectations of Tendermint ABCI and Cosmos SDK applications?

Reply:
More or less. Although it applies also to modules not maintained by the sdk team

Reply:
Gotcha. So then the ADR should describe how to define and enforce compatibility guarantees across the combinatorial sum of all supported modules provided by the SDK, at least. Like, if the SDK has x/bank/{v1,v2} and x/staking/{v1,v2,v3} then it must also have tests that ensure compatibility of apps which use

- x/bank/v1 + x/staking/v1
- x/bank/v1 + x/staking/v2
- x/bank/v1 + x/staking/v3
- x/bank/v2 + x/staking/v1
- x/bank/v2 + x/staking/v2
- x/bank/v2 + x/staking/v3 

and so on.

Reply:
Exactly

Reply:
So how you gonna do that? 🙃

Reply:
> So how you gonna do that? upside_down_face

Probably integration tests, and shell script generating go.mods with different combinations of module versions.

Reply:
> > So how you gonna do that? upside_down_face
> 
> Probably integration tests, and shell script generating go.mods with different combinations of module versions.

Possible! Excited to see the script that does that work. Should be quite a journey! And certainly no merge until that's passing, I suppose.

Reply:
If we do proper semver then we don't even need a shell script because we can build a single binary with bank/v1, bank/v2, etc. I'm thinking there would be a test runner that takes a list of versions for each module and runs the matrix.

Reply:
Sure! As long as e.g. bank/v1 always maps to the the same .proto definitions of all relevant types, invariant to Git SHA. If e.g. bank/v1 at SHA 0000001 uses bank.proto with one definition of Account, and and bank/v1 at SHA 0000002 uses bank.proto with a different definition of Account, then those two bank/v1 SHAs need to be tested independently! 😬

Reply:
That's one of the points this ADR addresses - the file descriptors for each module version need to be registered and tested independently

Reply:
Can you link to precisely what you mean when you say "file descriptors"?

Reply:
FileDescriptorProto: https://buf.build/beta/wellknowntypes/docs/main:google.protobuf#google.protobuf.FieldDescriptorProto